### PR TITLE
Add edit parameter to plumb_api()

### LINKED ISSUE #699: Add `edit` parameter to `plumb_api()`
After an extensive conversation with @schloerke, we've decided to add an `edit` parameter to the `plumb_api()` function. This parameter will open the `plumber.R` or `entrypoint.R` file associated with a specified Plumber API. This will enable Plumber users to quickly and easily view source code for Plumber APIs that are included in R packages.

The expected behavior of this feature is as follows:
```r
plumb_api("plumber", "01-append", edit = TRUE)
```
1. A message is presented to the user specifying the path of the directory containing the source code and a warning that if they want to make persistent changes to the code, they should copy it to a separate file.
2. The associated `plumber.R` file is opened for editing
3. If using RStudio, the files pane is directed to the directory containing the source code

There were a few alternative options discussed that were ultimately rejected:

- Copying the source directory to a temporary location and opening the files from there
  This idea was based on preventing users from accidentally changing the behavior of API examples included with installed packages. It was ultimately decided that the simplest solution is to just open the original source file and provide the user with a warning about permanently editing the files.

- Creating a separate `pr_view_source()` function.
  This idea was based around the private `$filename` value for Plumber routers. The idea was to capture that filename and to open the source file. This approach was rejected in favor of adding `edit` to `plumb_api()` for simplicity sake. Using `plumb_api()` instead of introducing a new function also restricts this functionality to APIs included in installed packages, which is a desired characteristic.

### PR DESCRIPTION
**Pull request by blairj09**
Closes #699

PR task list:
- [x] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`

A few discussion points:

- As it stands today, there is no way to programmatically set the RStudio Files pane to a specific location (see https://github.com/rstudio/rstudioapi/issues/204)
- At this point, the user is warned every time `edit = TRUE` is used. I actually like this pattern, b/c I don't anticipate this is a function that will be called with high frequency. However, we could consider only warning once per session.

### CODE REVIEW

File: R/plumb.R [hunk 1]
```diff
@@ -114,6 +116,23 @@ plumb_api <- function(package = NULL, name = NULL) {
...
+              call. = FALSE)
+    }
+
+    if (file.exists(file.path(api_dir, "entrypoint.R"))) {
+      file_loc <- file.path(api_dir, "entrypoint.R")
+    } else {
+      file_loc <- file.path(api_dir, "plumber.R")
+    }
+      file.edit(file_loc)
+      edit_warning(file_loc, package)
```
Comment (schloerke) [inline R/plumb.R:133]:
```suggestion
    file.edit(file_loc)
    edit_warning(file_loc, package)
```

File: tests/testthat/test-zzz-plumb_api.R [hunk 2]
```diff
@@ -57,6 +57,34 @@ test_that("errors are thrown", {
...
+
+  selected_api <- apis$package == "plumber" & apis$name == "01-append"
+
+  expect_output(plumb_api("plumber", "01-append", edit = TRUE),
+                file.path(apis[selected_api, "source_directory"], "plumber.R"))
+
+  selected_api <- apis$package == "plumber" & apis$name == "12-entrypoint"
+
+  expect_output(plumb_api("plumber", "12-entrypoint", edit = TRUE),
+                file.path(apis[selected_api, "source_directory"], "entrypoint.R"))
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:76]:
```suggestion
  expect_output(
  	plumb_api("plumber", "01-append", edit = TRUE),
    paste0(file.path(apis[selected_api, "source_directory"], "plumber.R"), " has been opened")
  )

  selected_api <- apis$package == "plumber" & apis$name == "12-entrypoint"

  expect_output(
    plumb_api("plumber", "12-entrypoint", edit = TRUE),
    paste0(file.path(apis[selected_api, "source_directory"], "entrypoint.R"), " has been opened")
  )
```

File: R/plumb.R [hunk 3]
```diff
@@ -114,6 +116,23 @@ plumb_api <- function(package = NULL, name = NULL) {
     stop("Could not find Plumber API for package '", package, "'  with name '", name, "'")
   }
 
+  api_dir <- apis[apis_sub, "source_directory"]
+
+  if (edit) {
+    edit_warning <- function(file_path, package) {
+      warning(file_path, " has been opened in the editor. Any changes saved to this file are permament until the ", package, " package is reinstalled. If you would like to make persistent changes, consider copying the contents of this file to a new file.",
+              call. = FALSE)
+    }
```
Comment (schloerke) [inline R/plumb.R:125]:
```suggestion
```

File: R/plumb.R [hunk 4]
```diff
@@ -114,6 +116,23 @@ plumb_api <- function(package = NULL, name = NULL) {
     stop("Could not find Plumber API for package '", package, "'  with name '", name, "'")
   }
 
+  api_dir <- apis[apis_sub, "source_directory"]
+
+  if (edit) {
```
Comment (schloerke) [inline R/plumb.R:121]:
```suggestion

  if (isTRUE(edit)) {
    api_dir <- apis[apis_sub, "source_directory"]
```

File: R/plumb.R [hunk 5]
```diff
@@ -91,10 +91,12 @@ plumb <- function(file = NULL, dir = ".") {
 #'
 #' @param package Package to inspect
 #' @param name Name of the package folder to [plumb()].
+#' @param edit Whether or not to open the API source code for viewing / editing
 #' @describeIn plumb_api [plumb()]s a package's Plumber API. Returns a [`Plumber`] router object
 #' @return A [`Plumber`] object. If either `package` or `name` is null, the appropriate [available_apis()] will be returned.
+#' @importFrom utils file.edit
```
Comment (schloerke) [inline R/plumb.R:97]: This is not needed for speed. Would rather call it directly.

```suggestion
```

Comment (schloerke) [reply R/plumb.R:97]: jk. I think this would prevent rstudio from shimming it

File: tests/testthat/test-zzz-plumb_api.R [hunk 6]
```diff
@@ -57,6 +57,34 @@ test_that("errors are thrown", {
...
+
+
+  # Reset editor
+  options(editor)
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:80]:
```suggestion
```

File: tests/testthat/test-zzz-plumb_api.R [hunk 7]
```diff
@@ -57,6 +57,34 @@ test_that("errors are thrown", {
   expect_error(available_apis("crayon"), "No Plumber APIs found for package")
 })
 
+test_that("edit opens correct file", {
+  # Redefine editor so that file.edit doesn't try to open a file
+  editor <- options(editor = function(name, file, title) {
+    print(file)
+  })
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:64]:
```suggestion
  orig_opts <- options(editor = function(name, file, title) {
    print(file)
  })
  on.exit(options(orig_opts), add = TRUE)
```

File: tests/testthat/test-zzz-plumb_api.R [hunk 8]
```diff
@@ -57,6 +57,34 @@ test_that("errors are thrown", {
...
+
+
+  # Reset editor
+  options(editor)
+})
+
+test_that("edit throws a warning", {
+  editor <- options(editor = function(name, file, title) NULL)
+  expect_warning(plumb_api("plumber", "01-append", edit = TRUE))
+  options(editor)
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:86]:
```suggestion
  orig_opts <- options(editor = function(name, file, title) NULL)
  on.exit(options(orig_opts), add = TRUE)
  expect_warning(plumb_api("plumber", "01-append", edit = TRUE))
```

**Review by schloerke — APPROVED** [review]

File: tests/testthat/test-zzz-plumb_api.R [hunk 9]
```diff
@@ -57,6 +57,35 @@ test_that("errors are thrown", {
   expect_error(available_apis("crayon"), "No Plumber APIs found for package")
 })
 
+test_that("edit opens correct file", {
+  # Redefine editor so that file.edit doesn't try to open a file
+  orig_opts <- options(editor = function(name, file, title) {
+    print(file)
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:63]:
```suggestion
    cat(file, "has been opened\n")
```

File: tests/testthat/test-zzz-plumb_api.R [hunk 10]
```diff
@@ -57,6 +57,35 @@ test_that("errors are thrown", {
   expect_error(available_apis("crayon"), "No Plumber APIs found for package")
 })
 
+test_that("edit opens correct file", {
+  # Redefine editor so that file.edit doesn't try to open a file
+  orig_opts <- options(editor = function(name, file, title) {
+    cat(file, "has been opened\n")
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:63]:
```suggestion
    cat(file, " test file attempted to open\n")
```

File: tests/testthat/test-zzz-plumb_api.R [hunk 11]
```diff
@@ -57,6 +57,35 @@ test_that("errors are thrown", {
...
+  expect_output(
+  	plumb_api("plumber", "01-append", edit = TRUE),
+    paste0(file.path(apis[selected_api, "source_directory"], "plumber.R"), " has been opened")
+  )
+
+  selected_api <- apis$package == "plumber" & apis$name == "12-entrypoint"
+
+  expect_output(
+    plumb_api("plumber", "12-entrypoint", edit = TRUE),
+    paste0(file.path(apis[selected_api, "source_directory"], "entrypoint.R"), " has been opened")
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:80]:
```suggestion
    paste0(file.path(apis[selected_api, "source_directory"], "entrypoint.R"), " test file attempted to open")
```

File: tests/testthat/test-zzz-plumb_api.R [hunk 12]
```diff
@@ -57,6 +57,35 @@ test_that("errors are thrown", {
...
+  expect_output(
+  	plumb_api("plumber", "01-append", edit = TRUE),
+    paste0(file.path(apis[selected_api, "source_directory"], "plumber.R"), " has been opened")
+  )
+
+  selected_api <- apis$package == "plumber" & apis$name == "12-entrypoint"
+
+  expect_output(
+    plumb_api("plumber", "12-entrypoint", edit = TRUE),
+    paste0(file.path(apis[selected_api, "source_directory"], "entrypoint.R"), " test file attempted to open")
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:80]:
```suggestion
    "entrypoint.R test file attempted to open",
    fixed = TRUE
```

File: tests/testthat/test-zzz-plumb_api.R [hunk 13]
```diff
@@ -57,6 +57,37 @@ test_that("errors are thrown", {
   expect_error(available_apis("crayon"), "No Plumber APIs found for package")
 })
 
+test_that("edit opens correct file", {
+  # Redefine editor so that file.edit doesn't try to open a file
+  orig_opts <- options(editor = function(name, file, title) {
+    cat(file, " test file attempted to open\n")
```
Comment (schloerke) [inline tests/testthat/test-zzz-plumb_api.R:63]: There should be a function called `cat0()`

```suggestion
    cat(file, " test file attempted to open\n", sep = "")
```

**Comment by blairj09** [comment]
Just FYI, GH version of rstudioapi has the ability to navigate the files pane now: https://github.com/rstudio/rstudioapi/pull/206

**Comment by schloerke** [comment]
But also requires a dev version of the IDE.

Will need to update this code once that IDE is shipped.

**Pull request merged**